### PR TITLE
refactor: implement reverse argument binding

### PR DESCRIPTION
**Pull request by jrandolf-google**
This PR implements reverse binding and refactors the binding code within Puppeteer.

Reverse argument bindings allow complex objects such as `Node`s to be passed from the client to Puppeteer (similar to how we can pass handles to the client). For now, we only install `Node` argument bindings for ARIA queries, but more can be installed as use-cases come up.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,17 +26,92 @@
...
-    "test:chrome:headful": "npm test -- --test-suite chrome-headful",
-    "test:chrome:new-headless": "npm test -- --test-suite chrome-new-headless",
-    "test:chrome:headless": "npm test -- --test-suite chrome-headless",
-    "test:chrome:bidi": "npm test -- --test-suite chrome-bidi",
-    "test:chrome": "run-s test:chrome:*",
-    "test:firefox:bidi": "npm test -- --test-suite firefox-bidi",
-    "test:firefox:headful": "npm test -- --test-suite firefox-headful",
-    "test:firefox:headless": "npm test -- --test-suite firefox-headless",
-    "test:firefox": "run-s test:firefox:*",
+    "test:chrome:headful": "wireit",
```
Comment (OrKoN) [inline package.json]: This change shows in multiple PRs it seems. It's better be moved in a separate PR as it does not seem to be directly related to the goal of this change.

**Comment by OrKoN** [comment]
What is a reverse argument binding and why do we need it?

**Comment by jrandolf** [comment]
> What is a reverse argument binding and why do we need it?

I've added a description within the PR now.

File: packages/puppeteer-core/src/common/Binding.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+export class Binding {
+  #name: string;
+  #fn: (...args: unknown[]) => unknown;
+  constructor(name: string, fn: (...args: unknown[]) => unknown) {
+    this.#name = name;
+    this.#fn = fn;
+  }
+
+  /**
+   * @param throwError - Whether to throw the error instead of sending it to the
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/Binding.ts]: outdated comment?

File: packages/puppeteer-core/src/common/Binding.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+              const callbacks = globalThis[name].callbacks;
+              callbacks.get(seq).resolve(result);
+              callbacks.delete(seq);
+            },
+            this.#name,
+            seq,
+            await this.#fn(...args)
+          );
+        } finally {
+          for (const [, handle] of properties) {
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/Binding.ts]: do we need to dispose sequentially?

File: packages/puppeteer-core/src/common/types.ts [hunk 4]
```diff
@@ -14,9 +14,19 @@
...
+import type {ElementHandle} from '../api/ElementHandle.js';
+import type {LazyArg} from './LazyArg.js';
+
+/**
+ * @internal
+ */
+export type BindingPayload = {
+  type: string;
+  name: string;
+  seq: number;
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/types.ts:27]: what does seq mean? Pls add a comment.

Comment (OrKoN) [reply packages/puppeteer-core/src/common/types.ts:27]: I guess is the id of a callback/promise?

File: packages/puppeteer-core/src/common/Binding.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  constructor(name: string, fn: (...args: unknown[]) => unknown) {
+    this.#name = name;
+    this.#fn = fn;
+  }
+
+  /**
+   * @param throwError - Whether to throw the error instead of sending it to the
+   * binding.
+   */
+  async call(
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/Binding.ts]: nit: perhaps the name of this method could be `onCall` or smth like this to indicate that the binding is not invoked here but it is the logic that happens after the page called the binding? Although, I could see how `call` could reflect that but I think it is also more confusing.

File: packages/puppeteer-core/src/common/Binding.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    context: ExecutionContext,
+    seq: number,
+    args: unknown[]
+  ): Promise<void> {
+    try {
+      // Getting non-trivial arguments.
+      const handles = await context.evaluateHandle(
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/Binding.ts]: so we do those extra calls on every binding call even if there were no non-primitive arguments?

Comment (jrandolf) [reply packages/puppeteer-core/src/common/Binding.ts]: Done.

File: packages/puppeteer-core/src/common/Binding.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+              const callbacks = globalThis[name].callbacks;
+              callbacks.get(seq).reject(error);
+              callbacks.delete(seq);
+            },
+            this.#name,
+            id,
+            error.message,
+            error.stack
+          )
+          .catch(() => {});
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/Binding.ts]: how often can it fail? let's handle it through the debugError helper, just in case if smth stops working and we need to debug. The same below.

**Review by OrKoN — APPROVED** [review]
lgtm % debugError suggestion
